### PR TITLE
Allow creating sockets from raw file descriptors

### PR DESCRIPTION
**Pull request by myfreeweb**
This is useful for [socket activation](https://github.com/myfreeweb/soad), sandboxes/fd-passing, etc.

### CODE REVIEW

**Comment by ysbaddaden** [comment]
Use case is valid, but CI is red.

**Comment by straight-shoota** [comment]
Needs fix and maybe a spec for the described behaviour.

**Comment by straight-shoota** [comment]
@myfreeweb Are you still working on this?

**Comment by myfreeweb** [comment]
I don't understand why that test is failing..

**Comment by straight-shoota** [comment]
@myfreeweb The added `TCPServer#initialize(fd)` shadows `TCPServer.new(port)` because both can be called with a single Int32 argument.

I'd suggest to make `fd` an required named argument: `def initialize(*, fd : Int32 ...` This should be changed uniquely in all fd constructors.

**Comment by straight-shoota** [comment]
It's a shame that such a change breaks only one mostly unrelated SSL spec 😢

**Comment by myfreeweb** [comment]
Passes now

**Comment by straight-shoota** [comment]
Awesome :+1: Could you add a doc to the now public `Socket#initialize` as well?

**Comment by myfreeweb** [comment]
done

File: src/socket.cr [hunk 1]
```diff
@@ -79,7 +79,8 @@ class Socket < IO
     end
   end
 
-  protected def initialize(@fd : Int32, @family, @type, @protocol = Protocol::IP, blocking = false)
+  # Creates a Socket from an existing socket file descriptor.
+  def initialize(@fd : Int32, @family, @type, @protocol = Protocol::IP, blocking = false)
```
Comment (asterite) [inline src/socket.cr]: Shouldn't this also use named arguments like the other `initialize` methods in subtypes?

Comment (myfreeweb) [reply src/socket.cr]: There's no conflict on `Socket`.. named arguments are only really necessary for TCP

Comment (asterite) [reply src/socket.cr]: I see. I meant for consistency, but it's not a big deal.

**Pull request merged**